### PR TITLE
[WIP] Improving Tx errors

### PR DESCRIPTION
**Pull request by ramtinms**
This PR proposes a few changes: 
- aggregates all tx errors into an error package (more visibility for users)
- breaks some of the more generic errors into more detailed ones and group some others into a single error (based on the type of action needed by user, nodes, ... )
- makes the road smoother for some of the slashing challenges (tx validation steps to be verified by exec nodes and verif nodes) 
- helps to handle errors with more clarity at FVM layer.

### CODE REVIEW

File: model/errors/transaction.go [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+)
+
+// TransactionValidationError captures a transaction validation error
+// A transaction having this error (in most cases) is rejected by access/collection nodes
+// and later in the pipeline be verified by execution and verification nodes.
+type TransactionValidationError interface {
+	// Code returns the code for this error
+	Code() uint32
+	// Error returns an string describing the details of the error
+	Error() string
```
Comment (jordanschalm) [inline model/errors/transaction.go:51]: I assume the purpose of this is to ensure these can also pass as basic Go errors, in which case I'd suggest we embed the `error` interface here to make that explicit.
```suggestion
	error
```

File: model/errors/transaction.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+// account doesn't have enough storage
+type InsufficientStorageError struct {
+	Capacity uint64
+	Used     uint64
+	Needed   uint64
+}
+
+func (e InsufficientStorageError) TxHash() flow.Identifier {
+	return e.TxHash
+}
```
Comment (jordanschalm) [inline model/errors/transaction.go:438]: The only way to actually call these `TxHash` methods would be to type assert to the structure type (ie. `InsufficientStorageError`) in which case you can just do `err.TxHash`. I think this scheme causes naming collisions as well (`err.TxHash` is ambiguous currently)

File: model/errors/transaction.go [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+	errCodeInsufficientTokenBalanceError         = 201
+	errCodeMaxGasExceededError                   = 202
+	errCodeMaxEventLimitExceededError            = 203
+	errCodeMaxLedgerIntractionLimitExceededError = 204
+)
+
+// TransactionValidationError captures a transaction validation error
+// A transaction having this error (in most cases) is rejected by access/collection nodes
+// and later in the pipeline be verified by execution and verification nodes.
+type TransactionValidationError interface {
```
Comment (jordanschalm) [inline model/errors/transaction.go:47]: 👍 I'm onboard with this approach. I wasn't sure, so made an example to confirm this still works with `errors.As`:
https://play.golang.org/p/EjLQ_jAbXpC

**Review by jordanschalm — COMMENTED** [review]
Nice work, I think this is a approach

**Comment by ramtinms** [comment]
This is deprecated, new adapted version is in this PR https://github.com/onflow/flow-go/pull/563. 
Thanks, @jordanschalm for all the comments I applied them to the new PR.

**Pull request closed without merging**